[master][CC] Update Websocket docs and Improve Deploy API docs (#5595)
wso2/docs-apim: en/docs/deploy-and-publish/deploy-on-gateway/choreo-connect/deploy-api/deploy-websocket-api-in-choreo-connect.md

# Deploying a WebSocket API in Choreo Connect

You can deploy a WebSocket API in the following ways depending on the Choreo Connect **mode** you have chosen.

|**Mode**         | **Method**    |
|--------------|-----------|
|[Choreo Connect with WSO2 API Manager as a Control Plane]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/concepts/apim-as-control-plane/)   | [Via WSO2 API Manager Publisher Portal](#via-wso2-api-manager-publisher-portal)  |
|[Choreo Connect as a Standalone Gateway]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/concepts/as-a-standalone-gateway/)  |[Via apictl for Standalone Mode](#via-apictl-for-standalone-mode) |

> !!! tip "Characteristics of Websocket APIs Deployed in Choreo Connect"

Reviewer: ```suggestion
!!! tip "Characteristics of WebSocket APIs Deployed in Choreo Connect"
```

    - The API is created based on the **Channels** defined in the AsyncAPI definition. They are also known as **Topics** in WSO2 API Manager, and have a slight similarity to **Resources** in a REST API.

>     - Once the API is exposed, although the gateway endpoint includes a basepath (with topic paths if defined), the **Connection Request** will be sent to the original backend endpoint you provided without any appended paths.

Reviewer: ```suggestion
    - Once the API is exposed, although the Gateway endpoint includes a basepath (with topic paths if defined), the **Connection Request** will be sent to the original backend endpoint you provided without any appended paths.
```

    - If you have a requirement to append different **paths** to the backend URL for different **Channels** (**Topics**), a URL-mapping can be added. The **Topics** that do not include a URL-mapping will have the usual behavior as mentioned above.

>     - Websocket APIs support OAuth 2.0 Security and each **Topic** can have its own local scope.

Author: ```suggestion
    - Websocket APIs support OAuth 2.0 Security and each **Topic** can have its own local scope.
    - If configured, Choreo Connect also capable of publishing analytics for websocket frames sent and received by the client.
```

>     - Websocket APIs support OAuth 2.0 Security and each **Topic** can have its own local scope.         

Reviewer: ```suggestion
    - WebSocket APIs support OAuth 2.0 Security and each **Topic** can have its own local scope.         
```

>     - If configured, Choreo Connect is also capable of publishing analytics for websocket frames sent and received by the client.         

Reviewer: ```suggestion
    - If configured, Choreo Connect is also capable of publishing analytics for WebSocket frames sent and received by the client.         
```

>     - Currently, the properties such as URL Mapping and Security are supported at the topic action level with a similarity to REST APIs. Therefore, in Websocket APIs these must be defined under **Publish** or **Subscribe**. Yet, from a Gateway point of view, since both Publish and Subscribe are initiated with a GET request that appears the same, if a **Topic** has properties defined under both Publish and Subscribe, the properties for **Publish** will be prioritized.

Reviewer: ```suggestion
    - Currently, the properties such as URL Mapping and Security are supported at the topic action level with similarity to REST APIs. Therefore, in WebSocket APIs, these must be defined under **Publish** or **Subscribe**. Yet, from a Gateway point of view, since both Publish and Subscribe are initiated with a GET request that appears the same, if a **Topic** has properties defined under both Publish and Subscribe, the properties for **Publish** will be prioritized.
```

## Via WSO2 API Manager Publisher Portal

> Follow the instructions below to deploy a WebSocket type Streaming API to Choreo Connect via the Publisher Portal in WSO2 API Manager. The Publisher portal provides the capability to conveniently add and update the following.

Reviewer: ```suggestion
Follow the instructions below to deploy a WebSocket type Streaming API to Choreo Connect via the Publisher Portal in WSO2 API Manager. The Publisher Portal provides the capability to conveniently add and update the following.
```

### Step 1 - Create a WebSocket API in API Manager

1. Create a WebSocket API by following the steps in [Create a WebSocket API]({{base_path}}/design/create-api/create-streaming-api/create-a-websocket-streaming-api/).

2. By default, a WebSocket API is protected by OAuth2 in Choreo Connect. Make sure to switch on security for the API via Publisher and **Save**.

    [![Security switched on]({{base_path}}/assets/img/design/create-api/streaming-api/streaming-api-security-on.png)]({{base_path}}/assets/img/design/create-api/streaming-api/streaming-api-security-on.png)

### Step 2 - Deploy and publish the API

1. Deploy the API in Choreo Connect by navigating to the **Deployments** page from the left menu. For more information, see [Deploy API]({{base_path}}/deploy-and-publish/deploy-on-gateway/deploy-api/deploy-an-api/).
2. Publish the API from the **Lifecycle** page.

### Step 3 - Generate an Access Token to invoke the API

!!! tip 

>     To generate a temporary test key to invoke the API follow the steps [here]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/security/generate-a-test-jwt/)

Reviewer: ```suggestion
    To generate a temporary test key to invoke the API, follow the steps [here]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/security/generate-a-test-jwt/)
```

1. Click **View in Dev Portal** at the top right corner to open Developer Portal in another browser tab.

2. Open **Applications** from the top menu and select **DefaultApplication** from the list.

3. Open the **Subscriptions** tab and subscribe your API to the **DefaultApplication**.

4. Open **APIs** from the top menu and select your API.

5. Open the **Subscriptions** tab from the left menu bar, click on **PROD KEYS**, and generate keys.

> 6. Open the **Try Out** tab and click **GET TEST KEY**. This will include the access token in the cURL command you generate in the section below

Reviewer: ```suggestion
6. Open the **Try Out** tab and click **GET TEST KEY**. This will include the access token in the cURL command you generate in the section below.
```

### Step 4 - Invoke the API using a WebSocket client

The WebSocket API exposed via Choreo Connect can be invoked using a WebSocket client.

1. Install wscat client.

2. From Developer Portal resource tab open the topic `/notifications` and generate the cURL command.

> 3. Open a Terminal and run the the cURL command.

Reviewer: ```suggestion
3. Open a command-line interface and run the cURL command.
```

4. Once connected, type anything you want. The websocket server you are connected to must echo back what you type.

!!! note
    Choreo Connect handles websocket connections via the same ports used for HTTP (9090) and HTTPS (9095).

## Via apictl for Standalone Mode

> The CLI tool **apictl** does not support initializing projects for Streaming APIs yet. However, you can download a Websocket API from a WSO2 API Manager instance and deploy the downloaded project in Choreo Connect. You do not need any special configurations in API-M or Choreo Connect for this. If you do not have a running API Manager instance already simply download the pack from the [official site](https://wso2.com/api-manager/) and start one locally.

Reviewer: ```suggestion
The CLI tool **apictl** does not support initializing projects for Streaming APIs yet. However, you can download a WebSocket API from a WSO2 API Manager instance and deploy the downloaded project in Choreo Connect. You do not need any special configurations in API-M or Choreo Connect for this. If you do not have a running API Manager instance already, simply download the pack from the [official site](https://wso2.com/api-manager/) and start one locally.
```

> Following are the steps to download, and deploy the API using apictl.

Reviewer: ```suggestion
The following are the steps to download and deploy the API using apictl.
```

### Step 1 - Create an API 

Create a WebSocket API following the exact steps in [Create a WebSocket API in API Manager](#step-1-create-a-websocket-api-in-api-manager). 

### Step 2 - Download the API

> After creating the API and making the necessary changes click **Download API** in the top right corner. The downloaded zip can be deployed via apictl using the next few commands.

Reviewer: ```suggestion
After creating the API and making the necessary changes, click **Download API** in the top right corner. The downloaded .zip file can be deployed via apictl using the next few commands.
```

|[Choreo Connect as a Standalone Gateway]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/concepts/as-a-standalone-gateway/)  |[Via apictl for Standalone Mode](#via-apictl-for-standalone-mode) |

!!! tip "Characteristics of Websocket APIs Deployed in Choreo Connect"
    - The API is created based on the **Channels** defined in the AsyncAPI definition. They are also known as **Topics** in WSO2 API Manager, and have a slight similarity to **Resources** in a REST API.
    - Once the API is exposed, although the gateway endpoint includes a basepath (with topic paths if defined), the **Connection Request** will be sent to the original backend endpoint you provided without any appended paths.
    - If you have a requirement to append different **paths** to the backend URL for different **Channels** (**Topics**), a URL-mapping can be added. The **Topics** that do not include a URL-mapping will have the usual behavior as mentioned above.

>     - Websocket APIs support OAuth 2.0 Security and each **Topic** can have its own local scope.
>     - If configured, Choreo Connect also capable of publishing analytics for websocket frames sent and received by the client.

Author: ```suggestion
    - Websocket APIs support OAuth 2.0 Security and each **Topic** can have its own local scope.         
    - If configured, Choreo Connect is also capable of publishing analytics for websocket frames sent and received by the client.         
```

    - Currently, the properties such as URL Mapping and Security are supported at the topic action level with a similarity to REST APIs. Therefore, in Websocket APIs these must be defined under **Publish** or **Subscribe**. Yet, from a Gateway point of view, since both Publish and Subscribe are initiated with a GET request that appears the same, if a **Topic** has properties defined under both Publish and Subscribe, the properties for **Publish** will be prioritized.

## Via WSO2 API Manager Publisher Portal

Follow the instructions below to deploy a WebSocket type Streaming API to Choreo Connect via the Publisher Portal in WSO2 API Manager. The Publisher portal provides the capability to conveniently add and update the following.

- The AsyncAPI definition
- Rate limiting policies
- Security scopes
- URL mappings

> 

Author: ```suggestion
```
